codegen/facelift/templates/IPCProxyAdapter.template.cpp

```diff
@@ -55,76 +51,132 @@
+            emitChangeSignal[QStringLiteral("ready")] = (previousIsReady != m_serviceReady);
+        }
     }
-    {% elif property.type.is_model %}
-    bool emit_{{property.name}}ChangeSignal = false;
-    if (isCompleteSnapshot) {
-        int {{property.name}}Size;
-        deserializeValue(msg, {{property.name}}Size);
-        m_{{property.name}}.beginResetModel();
-        m_{{property.name}}.reset({{property.name}}Size, std::bind(&ThisType::{{property.name}}Data, this, std::placeholders::_1));
-        m_{{property.name}}.endResetModel();
-        emit_{{property.name}}ChangeSignal = true;
+    for (const QString &propertyName: emitChangeSignal.keys()) {
+        {% for property in interface.properties %}
+        if (propertyName == QStringLiteral("{{property.name}}") && emitChangeSignal[propertyName]) {
+            emit {{property.name}}Changed();
+        }
+        {% endfor %}
+        if (propertyName == QStringLiteral("ready") && emitChangeSignal[propertyName]) {
+            emit readyChanged();
+        }
     }
-    {% else %}
-    const auto previous_{{property.name}}_Value = m_{{property.name}};
-    deserializeOptionalValue(msg, m_{{property.name}}, isCompleteSnapshot);
-    bool emit_{{property.name}}ChangeSignal = isCompleteSnapshot && ((previous_{{property.name}}_Value != m_{{property.name}}));
-    {% endif %}
-    {% endfor %}
-
-    bool emit_ReadyChangeSignal = deserializeReadyValue(msg, isCompleteSnapshot) && isCompleteSnapshot;
-
-    {% for property in interface.properties %}
-    if (emit_{{property.name}}ChangeSignal)
-        emit {{property.name}}Changed();
-    {% endfor %}
-
-    if (emit_ReadyChangeSignal)
-        emit readyChanged();
-
 }
 
-void {{className}}::deserializeSignal(InputIPCMessage &msg)
+void {{className}}::handleSignals(InputIPCMessage& msg)
 {
-    SignalID member;
-    deserializeValue(msg, member);
-
+    Q_UNUSED(msg)
     {% for event in interface.signals %}
-    if (member == SignalID::{{event}})
-    {
+    if (msg.member() == QStringLiteral("{{event}}")) {
+        QListIterator<QVariant> argumentsIterator(msg.arguments());
         {% for parameter in event.parameters %}
         {{parameter.interfaceCppType}} param_{{parameter.name}};
-        deserializeValue(msg, param_{{parameter.name}});
+        param_{{parameter.name}} = (argumentsIterator.hasNext() ? castFromQVariant<{{parameter.interfaceCppType}}>(argumentsIterator.next()):{% if not parameter.type.is_interface %}{{parameter.interfaceCppType}}(){% else %}nullptr{% endif %});
         {% endfor %}
         emit {{event}}(
         {%- set comma = joiner(", ") -%}
         {%- for parameter in event.parameters -%}
             {{ comma() }}param_{{parameter.name}}
         {%- endfor -%}  );
-    } else
+    }
     {% endfor %}
+
+    {% if interface.hasModelProperty %}
+    this->onModelUpdateEvent(msg);
+    {% endif %}
+}
+
+const QList<QString>& {{className}}::getSignals() const
```

Comment:
this is relevant to the DBus layer only.

Reply:
I consider moving the property to anther layer over-complicating the design with no apparent benefits.

Reply:
yes, could be but there are other examples (like signatureHelper) inside Adapter layer which actually belong to dbus layer, I suggest to create an issue for that so that all of them can be addressed together.